code/osapi/osapi.cpp

```diff
@@ -314,142 +336,124 @@ void os_deinit()
 	SDL_Quit();
 }
 
-extern SCP_map<int, int> SDLtoFS2;
-extern void joy_set_button_state(int button, int state);
-extern void joy_set_hat_state(int position);
+void debug_int3(char *file, int line)
+{
+	mprintf(("Int3(): From %s at line %d\n", file, line));
 
-void os_poll()
+	gr_activate(0);
+
+	mprintf(("%s\n", dump_stacktrace().c_str()));
+
+#ifndef NDEBUG
+	SDL_TriggerBreakpoint();
+#endif
+
+	gr_activate(1);
+	// we have to call os_deinit() before abort() so we make sure that SDL gets
+	// closed out and we don't lose video/input control
+	os_deinit();
+
+	abort();
+}
+
+namespace os
 {
-	SDL_Event event;
+	namespace events
+	{
+		namespace
+		{
+			ListenerIdentifier nextListenerIdentifier;
 
-	while (SDL_PollEvent(&event)) {
-		switch (event.type) {
-		case SDL_WINDOWEVENT: {
-			if (event.window.windowID == SDL_GetWindowID(os_get_window())) {
-				switch (event.window.event) {
-				case SDL_WINDOWEVENT_MINIMIZED:
-					case SDL_WINDOWEVENT_FOCUS_LOST:
-					{
-						if (fAppActive) {
-							if (!Cmdline_no_unfocus_pause) {
-								game_pause();
-							}
-							
-							fAppActive = false;
-						}
-						break;
-					}
-					case SDL_WINDOWEVENT_MAXIMIZED:
-					case SDL_WINDOWEVENT_RESTORED:
-					case SDL_WINDOWEVENT_FOCUS_GAINED:
-					{
-						if (!fAppActive) {
-							if (!Cmdline_no_unfocus_pause) {
-								game_unpause();
-							}
-							
-							fAppActive = true;
-						}
-						break;
-					}
-					case SDL_WINDOWEVENT_CLOSE:
-						gameseq_post_event(GS_EVENT_QUIT_GAME);
-						break;
+			template<class E>
+			struct enum_hash
+			{
+				static_assert(std::is_enum<E>::value, "Type is not an enum!");
+
+			public:
+				size_t operator()(const E&e) const {
+					return std::hash<size_t>()(static_cast<size_t>(e));
 				}
-			}
-			
-			gr_activate(fAppActive);
-			
-			break;
-		}
+			};
 
-		case SDL_SYSWMEVENT:
-#ifdef WIN32
-#ifdef FS2_VOICER
-			switch(event.syswm.msg->msg.win.msg)
+			struct EventListenerData
 			{
-			case WM_RECOEVENT:
-				if ( Game_mode & GM_IN_MISSION && Cmdline_voice_recognition)
+				ListenerIdentifier identifier;
+				Listener listener;
+				int weight;
+
+				bool operator<(const EventListenerData& other) const
 				{
-					VOICEREC_process_event( event.syswm.msg->msg.win.hwnd );
+					return weight < other.weight;
 				}
-				break;
-			default:
-				break;
-			}
-#endif // FS2_VOICER
-#endif // WIN32
-			break;
+			};
 
-		case SDL_KEYDOWN:
-			if (SDLtoFS2[event.key.keysym.scancode]) {
-				key_mark(SDLtoFS2[event.key.keysym.scancode], 1, 0);
-			}
-			break;
+			SCP_unordered_map<SDL_EventType, SCP_vector<EventListenerData>, enum_hash<SDL_EventType>> eventListeners;
```

Comment:
The impact should only be a bit higher than the switch-case variant. However, I could try using a vector with `std::sort` and `std::lower_bound`. That would trade runtime-complexity for better cache usage.
Honestly, I don't think it's worth the effort. The number of elements in the map will be pretty small so a map will still be pretty efficient.